### PR TITLE
[mlir][ArmSME] Add optional mask operand to tile_store

### PR DESCRIPTION
**Pull request by c-rhodes**

### CODE REVIEW

File: mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td [hunk 1]
```diff
@@ -60,6 +60,12 @@ def TileElementWidthMatchesTileID : TypesMatchWith<
                   "::llvm::cast<VectorType>($_self).getElementType())"
                   ".getWidth())">;
 
+class HasMatchingMaskTypeConstraint<string vector, string mask> :
+  OptionalTypesMatchWith<
+    mask # " has i1 element type and same shape as " # vector,
+    vector, mask,
+    "::llvm::cast<mlir::VectorType>($_self).cloneWith({}, IntegerType::get($_ctxt, 1))">;
```
Comment (c-rhodes) [inline mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td:67]: @MacDue I've based on the constraint you're adding in #69604 but parameterized it

Comment (MacDue) [reply mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td:67]: You should be able to rebase and de-duplicate this now :)

Comment (c-rhodes) [reply mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td:67]: Done :)

File: mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td [hunk 2]
```diff
@@ -356,6 +358,11 @@ def TileStoreOp : ArmSME_Op<"tile_store"> {
     rank 2 with dynamic dimensions, since the operation is scalable, and the
     element type must be a scalar that matches the element type of the result.
 
+    An optional SSA value `mask` may be specified to mask out elements written
+    to the MemRef. The `mask` type is an `i1` vector of the same shape as the
+    vector type that matches how elements are written into the MemRef. Elements
+    whose corresponding mask element is `0` are masked out.
```
Comment (MacDue) [inline mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td:364]: nit: This seems a little too verbose to me:
```suggestion
    An optional `mask` may be provided, the shape of the mask corresponds to the `tile`,
    and selects which elements of the tile will be stored. 
```

File: mlir/test/Dialect/ArmSME/vector-ops-to-sme.mlir [hunk 3]
```diff
@@ -315,6 +315,20 @@ func.func @transfer_write_2d_f64(%vector : vector<[2]x[2]xf64>, %dest : memref<?
 
 // -----
 
+// CHECK-LABEL: func.func @transfer_write_2d_with_mask_f64(
+// CHECK-SAME:                                             %[[VECTOR:.*]]: vector<[2]x[2]xf64>,
+// CHECK-SAME:                                             %[[DEST:.*]]: memref<?x?xf64>,
+// CHECK-SAME:                                             %[[MASK:.*]]: vector<[2]x[2]xi1>) {
+// CHECK:         %[[C0:.*]] = arith.constant 0 : index
+// CHECK:         arm_sme.tile_store %[[VECTOR]], %[[DEST]]{{\[}}%[[C0]], %[[C0]]], {{.*}} : memref<?x?xf64>, vector<[2]x[2]xf64>
```
Comment (MacDue) [inline mlir/test/Dialect/ArmSME/vector-ops-to-sme.mlir:323]: Should this check `MASK` is used too?

**Review by MacDue — APPROVED** [review]
A couple of little comments:

**Pull request merged**